### PR TITLE
[functorch][dim] Refactor test_split、test_contrants to use instantiate_device_type_tests for multi-device compatibility;Refactor test_spectral_ops to split CUDA-only tests into standalone TestCUFFT class 

### PR DESCRIPTION
**Pull request by ltllt1**
### Summary
1. test_split.py: Refactor test_split test cases with `instantiate_device_type_tests`, unify multi-device test logic.
Only modify test file, no core runtime code changes, original CPU/CUDA test behaviors are fully preserved.
2. test_constraints.py: Refactor test_constraints.py to use TestCase and instantiate_device_type_tests, unifying multi-device testing logic and aligning with PyTorch's modern testing standards.
3. test_spectral.py: split CUDA-only tests into standalone TestCUFFT class

### key changes
1. Replaced manual device checks and hardcoded allocations with PyTorch's native, device-agnostic testing framework and tensor APIs.
2. Migrated legacy pytest parameterization and manual is_cuda conditional branches to PyTorch's native TestCase, instantiate_device_type_tests, and device-agnostic tensor APIs.
3. Extract all CUDA-specific cufft related test methods out of the main TestFFT class into a new independent TestCUFFT test class.

### Self-check
- [x] Local lint check passed
- [x] No breaking changes for existing devices
- [x] PyTorch CLA signed

### CODE REVIEW

File: test/functorch/dim/test_split.py [hunk 1]
```diff
@@ -320,25 +323,10 @@ def test_large_tensor_split(self):
             self.assertEqual(part.order(x, split_dims[i], z).shape, (10, 5, 20))
             self.assertEqual(split_dims[i].size, 5)
 
-    def test_device_handling(self):
+    def test_device_handling(self, device):
```
Comment (xuyun44) [inline test/functorch/dim/test_split.py]: there is a problem, origin testing both cuda(if available) and cpu. 
new version test both?

Comment (ltllt1) [reply test/functorch/dim/test_split.py]: I have tested the new version on local NPU environment.
It can be seen from the logs that the sample runs on both NPU and CPU at the same time.
Logs are attached below

test_split_with_requires_grad_cpu (__main__.TestSplitCPU)
Test split with tensors that require gradients. ... ok
test_device_handling_npu (__main__.TestSplitPRIVATEUSE1)
Test split behavior with different devices. ... [W716 03:32:56.386238736 NPUCachingAllocator.cpp:174] Warning: The current CANN and Soc versions require processing for 32 padding size, with memory allocation. (function operator())
ok
test_dim_object_split_all_bound_npu (__main__.TestSplitPRIVATEUSE1)
Test split with all Dim objects bound to specific sizes. ... ok
test_dim_object_split_mixed_bound_unbound_npu (__main__.TestSplitPRIVATEUSE1)
Test split with mix of bound and unbound Dim objects. ... ok

File: test/test_spectral_ops.py [hunk 2]
```diff
@@ -1686,12 +1560,141 @@ def test(self, device):
 
     setattr(TestFFTDocExamples, 'test_' + doc_test.name, skipCPUIfNoFFT(test))
 
+class TestCUFFT(TestCase):
+    @deviceCountAtLeast(1)
+    @onlyCUDA
```
Comment (KarhouTam) [inline test/test_spectral_ops.py]: @onlyCUDA is not needed in a test class that instantiated only for CUDA.

File: test/test_spectral_ops.py [hunk 3]
```diff
@@ -1686,12 +1560,141 @@ def test(self, device):
...
+
+        X = torch.stft(x, n_fft=100, window=window)
+
+        with self.assertRaisesRegex(RuntimeError, "istft input and window must be on the same device"):
+            torch.istft(X, n_fft=100, window=window.to(device))
+
+        with self.assertRaisesRegex(RuntimeError, "istft input and window must be on the same device"):
+            torch.istft(x.to(device), n_fft=100, window=window)
+
+    @onlyNativeDeviceTypes
```
Comment (KarhouTam) [inline test/test_spectral_ops.py]: ditto

File: test/test_spectral_ops.py [hunk 4]
```diff
@@ -1686,12 +1560,141 @@ def test(self, device):
 
     setattr(TestFFTDocExamples, 'test_' + doc_test.name, skipCPUIfNoFFT(test))
 
+class TestCUFFT(TestCase):
```
Comment (KarhouTam) [inline test/test_spectral_ops.py]: Lack of attribute: `hw_classification = HardwareClassification.CUDA`

File: test/functorch/dim/test_split.py [hunk 5]
```diff
@@ -320,25 +323,10 @@ def test_large_tensor_split(self):
...
-            t = cuda_tensor[x, y, z]
-
-            d1, d2 = Dim("d1", 4), Dim("d2", 8)
-            result = t.split([d1, d2], dim=y)
-
-            for i, part in enumerate(result):
-                ordered = part.order(x, d1 if i == 0 else d2, z)
-                self.assertEqual(ordered.device.type, "cuda")
-                self.assertEqual(ordered.shape[0], 3)
-                self.assertEqual(ordered.shape[2], 5)
```
Comment (KarhouTam) [inline test/functorch/dim/test_split.py:338]: Why this part removed?

File: test/functorch/dim/test_split.py
Comment (KarhouTam) [inline test/functorch/dim/test_split.py:1]: I wonder whether TestSplit should be tested on various devices or not. Seems it's device-agnostic and should have `HardwareClassification.GENERIC` tag. I think the only test case (`test_device_handling`) should be placed in another test class with `HardwareClassification.ACCELERATOR` tag.

File: test/distributions/test_constraints.py [hunk 6]
```diff
@@ -82,110 +84,76 @@
...
-                [
-                    f"Failed to biject_to({constraint})",
-                    f"x = {x}",
-                    f"biject_to(...)(x) = {y}",
-                ]
+    return constraint_fn(
+        *(torch.tensor(x, dtype=torch.double, device=device) if isinstance(x, list) else x for x in args)
+    )
+
+class TestConstraints(TestCase):
```
Comment (KarhouTam) [inline test/distributions/test_constraints.py:102]: Lack of `hw_classification = HardwareClassification.ACCELERATOR`

**Review by KarhouTam — CHANGES_REQUESTED** [review]
Thanks for your contribution. I've left some comments. PTAL. Also, personally, I suggest you to split this PR into 3 and each focuses on only one test file, which helps review.

File: test/test_spectral_ops.py (hunk 2 above)
Comment (ltllt1) [reply test/test_spectral_ops.py]: Got it, thanks for the reminder. I have remove the redundant @onlyCUDA decorator.

File: test/test_spectral_ops.py (hunk 3 above)
Comment (ltllt1) [reply test/test_spectral_ops.py]: Got it, thanks for the reminder. I have remove the redundant @onlyNativeDeviceTypesdecorator.

File: test/test_spectral_ops.py (hunk 4 above)
Comment (ltllt1) [reply test/test_spectral_ops.py]: Thanks for the catch. I've added hw_classification = HardwareClassification.CUDA.

File: test/functorch/dim/test_split.py
Comment (ltllt1) [reply test/functorch/dim/test_split.py:1]: Thanks for the suggestion. I've extracted test_device_handling into a standalone test class tagged with HardwareClassification.ACCELERATOR. The original TestSplit class remains device-agnostic.

File: test/distributions/test_constraints.py (hunk 6 above)
Comment (ltllt1) [reply test/distributions/test_constraints.py:102]: Thanks for the catch. I've added hw_classification = HardwareClassification.ACCELERATOR.

File: test/functorch/dim/test_split.py (hunk 5 above)
Comment (ltllt1) [reply test/functorch/dim/test_split.py:338]: Sorry, this part got removed inadvertently while resolving merge conflicts, and I’ve added it back now.

File: test/functorch/dim/test_split.py [hunk 7]
```diff
@@ -463,6 +438,39 @@ def test_plain_tensor_regular_split_still_works(self):
             self.assertTrue(isinstance(part, torch.Tensor))
             self.assertFalse(hasattr(part, "dims"))  # Should be regular tensor
 
+class TestSplitDevice(TestCase):
+    hw_classification = HardwareClassification.ACCELERATOR
+    def test_device_handling(self):
+        """Test split behavior with different devices."""
+        if torch.cuda.is_available():
```
Comment (xuyun44) [inline test/functorch/dim/test_split.py]: may need generalization
cuda -> accelerator

File: test/functorch/dim/test_split.py [hunk 8]
```diff
@@ -463,6 +451,27 @@ def test_plain_tensor_regular_split_still_works(self):
...
+            d1, d2 = Dim("d1", 4), Dim("d2", 8)
+            result = t.split([d1, d2], dim=y)
+
+            for i, part in enumerate(result):
+                ordered = part.order(x, d1 if i == 0 else d2, z)
+                self.assertEqual(ordered.device.type, device.type)
+                self.assertEqual(ordered.shape[0], 3)
+                self.assertEqual(ordered.shape[2], 5)
+
+instantiate_device_type_tests(TestSplit, globals())
```
Comment (KarhouTam) [inline test/functorch/dim/test_split.py]: Test cases in TestSplit don't have `device` arg, so this line of code is meaningless.

**Review by KarhouTam — APPROVED** [review]
LGTM. Thanks for your contribution!
cc @fffrog

**Review by xuyun44 — APPROVED** [review]

File: test/functorch/dim/test_split.py [hunk 9]
```diff
@@ -464,5 +451,28 @@ def test_plain_tensor_regular_split_still_works(self):
             self.assertFalse(hasattr(part, "dims"))  # Should be regular tensor
 
 
+class TestSplitDevice(TestCase):
+    hw_classification = HardwareClassification.ACCELERATOR
+
+    def test_device_handling(self, device):
+        """Test split behavior with different devices."""
+        if torch.accelerator.is_available():
```
Comment (fffrog) [inline test/functorch/dim/test_split.py]: Please delete this line, which is redundant.

File: test/functorch/dim/test_split.py [hunk 10]
```diff
@@ -464,5 +451,28 @@ def test_plain_tensor_regular_split_still_works(self):
...
+            result = t.split([d1, d2], dim=y)
+
+            for i, part in enumerate(result):
+                ordered = part.order(x, d1 if i == 0 else d2, z)
+                self.assertEqual(ordered.device, torch.device(device))
+                self.assertEqual(ordered.shape[0], 3)
+                self.assertEqual(ordered.shape[2], 5)
+
+
+instantiate_device_type_tests(TestSplitDevice, globals())
```
Comment (fffrog) [inline test/functorch/dim/test_split.py]: test_device_handling for CPU side have been added into `TestSplit`, so we should except_for="cpu" here.

File: test/test_spectral_ops.py [hunk 11]
```diff
@@ -12,7 +12,7 @@
```
Comment (fffrog) [inline test/test_spectral_ops.py:12]: This file requires a lot of changes, so would you please refactor it in a separate PR?

File: test/test_spectral_ops.py [hunk 12]
```diff
@@ -12,7 +12,7 @@
 
 from torch.testing._internal.common_utils import \
     (TestCase, run_tests, TEST_NUMPY, TEST_LIBROSA, requires_mkl, first_sample, TEST_WITH_ROCM,
-     make_tensor, skipIfTorchDynamo)
+     make_tensor, skipIfTorchDynamo, HardwareClassification)
 from torch.testing._internal.common_device_type import \
     (instantiate_device_type_tests, ops, dtypes, onlyNativeDeviceTypes,
      skipCPUIfNoFFT, deviceCountAtLeast, onlyCUDA, onlyOn, OpDTypes, toleranceOverride, tol,
```
Comment (fffrog) [inline test/test_spectral_ops.py:18]: Both `onlyOn` and `onlyCUDA` should be remoevd here.

**Review by fffrog — CHANGES_REQUESTED** [review]

File: test/functorch/dim/test_split.py (hunk 7 above)
Comment (ltllt1) [reply test/functorch/dim/test_split.py]: Thanks for the review! I've updated the implementation to use accelerator instead of cuda to make it more general and portable across different backends. PTAL

File: test/functorch/dim/test_split.py (hunk 9 above)
Comment (ltllt1) [reply test/functorch/dim/test_split.py]: Done. Removed the redundant line. Thanks!

File: test/functorch/dim/test_split.py (hunk 10 above)
Comment (ltllt1) [reply test/functorch/dim/test_split.py]: Done! Added except_for="cpu" since CPU-side tests are already covered in TestSplit. Thanks for the heads-up!

File: test/test_spectral_ops.py (hunk 11 above)
Comment (ltllt1) [reply test/test_spectral_ops.py:12]: Done! I've reverted the changes to this file in this PR and will handle the refactoring in a separate follow-up PR. Thanks!

File: test/test_spectral_ops.py (hunk 12 above)
Comment (ltllt1) [reply test/test_spectral_ops.py:18]: I've reverted the changes to this file in this PR and will handle the refactoring in a separate follow-up PR. Thanks!

File: test/functorch/dim/test_split.py [hunk 13]
```diff
@@ -321,22 +324,6 @@ def test_large_tensor_split(self):
...
-
-            d1, d2 = Dim("d1", 4), Dim("d2", 8)
-            result = t.split([d1, d2], dim=y)
-
-            for i, part in enumerate(result):
-                ordered = part.order(x, d1 if i == 0 else d2, z)
-                self.assertEqual(ordered.device.type, "cuda")
-                self.assertEqual(ordered.shape[0], 3)
-                self.assertEqual(ordered.shape[2], 5)
-
```
Comment (fffrog) [inline test/functorch/dim/test_split.py:339]: I think we can delete whole testcase `test_device_handling` here, because the CPU side is absolutely same as the CUDA side.

File: test/functorch/dim/test_split.py [hunk 14]
```diff
@@ -464,5 +451,27 @@ def test_plain_tensor_regular_split_still_works(self):
...
+        result = t.split([d1, d2], dim=y)
+
+        for i, part in enumerate(result):
+            ordered = part.order(x, d1 if i == 0 else d2, z)
+            self.assertEqual(ordered.device, torch.device(device))
+            self.assertEqual(ordered.shape[0], 3)
+            self.assertEqual(ordered.shape[2], 5)
+
+
+instantiate_device_type_tests(TestSplitDevice, globals(), except_for="cpu")
```
Comment (fffrog) [inline test/functorch/dim/test_split.py]: Remove except_for="cpu" please, we need to enable CPU side here.

**Review by fffrog — DISMISSED** [review]
SGTM, thank you for your contribution.

File: test/functorch/dim/test_split.py (hunk 14 above)
Comment (xuyun44) [reply test/functorch/dim/test_split.py]: if Remove except_for="cpu".
is hw_classification = HardwareClassification.ACCELERATOR an suitable label?

**Review by xuyun44 — APPROVED** [review]

File: test/functorch/dim/test_split.py (hunk 13 above)
Comment (ltllt1) [reply test/functorch/dim/test_split.py:339]: Done. Removed the test_device_handling. Thanks!

File: test/functorch/dim/test_split.py (hunk 14 above)
Comment (ltllt1) [reply test/functorch/dim/test_split.py]: Done. Removed except_for="cpu". Thanks!

**Comment by can-gaa-hou** [comment]
@claude review this pr with pr review skill

File: test/functorch/dim/test_split.py [hunk 15]
```diff
@@ -3,14 +3,17 @@
...
 from torch.testing._internal.common_utils import (
+    HardwareClassification,
     run_tests,
     TEST_WITH_TORCHDYNAMO,
     TestCase,
 )
 
 
 class TestSplit(TestCase):
+    hw_classification = HardwareClassification.GENERIC
```
Comment (mansiag05) [inline test/functorch/dim/test_split.py]: [nit]: The convention typically puts the docstring immediately after the class line. Could you please place _hw_classification = HardwareClassification.GENERIC_ after it.

Comment (ltllt1) [reply test/functorch/dim/test_split.py]: Done — moved hw_classification = HardwareClassification.GENERIC to after the class docstring in test/functorch/dim/test_split.py. Thanks for pointing it out!

**Review by mansiag05 — APPROVED** [review]
